File: Scala/src/main/scala/scripts/mixing/example/SubStateSetRepresentation.scala
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *\
*                   Simulation with reactive streams                     *
\* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

package scripts.mixing.example

import org.jzy3d.colors.Color
import tools.Plotting
import utils.ScriptBase


/* Sub-state representation of mixing problem
 *   onenote:https://d.docs.live.net/8e55450f976ac566/Notes/AI/Статьи.one#SWRS%20v0.2.5%20Basic%20modeling&
 *   section-id={40964F6B-F1E6-40E5-93D6-D7D464B4D57F}&page-id={AFB5685E-E61D-4707-B829-BABB190BC41D}&
 *   object-id={8EF68737-7BF8-08E1-0BCA-7F9EFBD51B94}&3E
 * Created 28.04.2018 author CAB */

object SubStateSetRepresentation extends ScriptBase with Plotting { import math._
  println(
    """ #### Sub-state set representation ####
      | X = [t]
      | Y = [ω_1, ω_2̂]
      | G = [v_1, v_2, q_1, q_2, q_3, q_4, ω_3]
    """.stripMargin)
  //Definitions
  case class valX(t: D)
  case class valY(ω_1: D, ω_2: D)
  trait domS
  case class valS1(ω_1: D) extends domS
  case class valS2(ω_2: D) extends domS
  trait domS_X{ val X: valX; val S: domS; val j: Int }
  case class S_X_1(X: valX, S: valS1, j: Int = 1) extends domS_X
  case class S_X_2(X: valX, S: valS2, j: Int = 2) extends domS_X
  //Functional model
  def F(X: valX): valY = {
    val q = sqrt(105.0)
    val em = ((q - 15.0) * X.t) / 16.0
    val ep = -(((q + 15.0) * X.t) / 16.0)
    valY(
      ω_1 = ((13.0 * exp(em) * q) / 21.0) - ((13.0 * exp(ep) * q) / 21.0) - (5.0 * exp(em)) - (5.0 * exp(ep)) + 10.0,
      ω_2 = -((5.0 * exp(em) * q) / 21.0) + ((5.0 * exp(ep) * q) / 21.0) + (5.0 * exp(em)) + (5.0 * exp(ep)) + 10.0)}
  //Generating of set of states
  val S_X_G: Set[domS_X] = (0 to 100).toSet[Int].flatMap{ i ⇒
    val X = valX(t = i / 10.0)
    val Y = F(X)
    Set(S_X_1(X, valS1(Y.ω_1)), S_X_2(X, valS2(Y.ω_2)))}
  println(s"Generated sub-states (|S_X_G| = ${S_X_G.size}):")
  S_X_G.foreach(S_X_j ⇒ println("    " + S_X_j))
  //Simulations
  def simulation(setX: Vector[valX]): Vector[valY] =setX.map{ valX ⇒
    //Collect all ub-state with same key
    val setS_valX = S_X_G.filter(_.X == valX)
    assume(setS_valX.size == 2, "Should be 2 sub-states for each valX key")
    //Build Y value
    valY(
      ω_1 = setS_valX.filter(_.j == 1).head.asInstanceOf[S_X_1].S.ω_1,
      ω_2 = setS_valX.filter(_.j == 2).head.asInstanceOf[S_X_2].S.ω_2)}
  //Run simulation in [0, 5]
  val setX: Vector[valX] = (0 to 50).map(i ⇒ valX(t = i / 10.0)).toVector
  val setY: Vector[valY] = simulation(setX)
  //Plot
  val (ωs1, ωs2) = setY.map(valY ⇒ (valY.ω_1, valY.ω_2)).unzip
  MultiYPlot2D(setX.map(_.t), Seq((ωs1, Color.RED), (ωs2, Color.GREEN))).show()}

File: Python/scripts/mixing_example/sub_state_set_representation.py
#!/usr/bin/env python

"""Sub-state set representation of mixing problem with building by function set model
onenote:https://d.docs.live.net/8e55450f976ac566/Notes/AI/Статьи.one#SWRS%20v0.2.5%20Basic%20modeling&
    section-id={40964F6B-F1E6-40E5-93D6-D7D464B4D57F}&page-id={AFB5685E-E61D-4707-B829-BABB190BC41D}&
    object-id={489EDFFB-40E0-073C-3F49-7F005EDECE76}&64
Created 27.04.2018 author CAB
"""

import math as m
import numpy as np
import matplotlib.pyplot as plt


# Script init
print(""" 
#### Sub-state set representation ####
X = [t]
Y = [ω_1, ω_2̂]
G = [v_1, v_2, q_1, q_2, q_3, q_4, ω_3]
""")

# Definitions
class 𝔛_:
    def __init__(self, t):
        self.t = t
    def __repr__(self):
        return f"𝔛 = [t = {self.t}]"
class 𝔜_:
    def __init__(self, ω_1, ω_2):
        self.ω_1 = ω_1
        self.ω_2 = ω_2
    def __repr__(self):
        return f"𝔜 = [ω_1 = {self.ω_1}, ω_2 = {self.ω_2}]"
class 𝔖𝔛_q_:  # Sub-state, 𝔖 ⊆ 𝔜 is sub-stat value, 𝔛 ∈ 𝕏 is key values set, q ∈ ℕ is sub-sate index
    def __init__(self, 𝔛, 𝔜, q):
        self.t = 𝔛.t
        if q == 1:
            self.ω_1 = 𝔜.ω_1
        elif q == 2:
            self.ω_2 = 𝔜.ω_2
        else:
            assert False, "index j can be only  1 or 2"
        self.q = q
    def __repr__(self):
        if self.q == 1:
            return f"𝔖^𝔛_q = 𝔖=[ω_1={self.ω_1}]^𝔛=[t={self.t}]_q=1"
        else:
            return f"𝔖^𝔛_q = 𝔖=[ω_2={self.ω_2}]^𝔛=[t={self.t}]_q=2"

# Function set model
def F(X):
    q = m.sqrt(105.0)
    em = m.exp(((q - 15.0) * X.t) / 16.0)
    ep = m.exp(-(((q + 15.0) * X.t) / 16.0))
    return 𝔜_(
        ω_1 = ((13.0 * em * q) / 21.0) - ((13.0 * ep * q) / 21.0) - (5.0 * em) - (5.0 * ep) + 10.0,
        ω_2 = -((5.0 * em * q) / 21.0) + ((5.0 * ep * q) / 21.0) + (5.0 * em) + (5.0 * ep) + 10.0)

# Generating of set of sub-states
set𝔖X𝔈 = []
for 𝔛 in [𝔛_(t) for t in np.round(np.arange(-15.0, 15.0, 0.1), 4)]:
    𝔜 = F(𝔛)
    set𝔖X𝔈.append(𝔖𝔛_q_(𝔛, 𝔜, q=1))
    set𝔖X𝔈.append(𝔖𝔛_q_(𝔛, 𝔜, q=2))
print("Generated set 𝔖^X|𝔈:")
for 𝔖𝔛_q in set𝔖X𝔈: print("    "+ str(𝔖𝔛_q))


# Simulation function
def simulation(set𝔛):
    set𝔜 = []
    for 𝔛 in set𝔛:
        𝔖𝔛_1 = None
        𝔖𝔛_2 = None
        for 𝔖𝔛_q in set𝔖X𝔈:
            if 𝔖𝔛_q.t == 𝔛.t and 𝔖𝔛_q.q == 1: 𝔖𝔛_1 = 𝔖𝔛_q
            if 𝔖𝔛_q.t == 𝔛.t and 𝔖𝔛_q.q == 2: 𝔖𝔛_2 = 𝔖𝔛_q
        set𝔜.append(𝔜_(𝔖𝔛_1.ω_1, 𝔖𝔛_2.ω_2))
    return set𝔜

# Run simulation
set𝔛 = np.vectorize(lambda t: 𝔛_(t))(np.round(np.arange(0.0, 10.1, 0.1), 4))
set𝔜 = simulation(set𝔛)

# Print result
print("Simulation result (𝔛 -> 𝔜): ")
for [𝔛, 𝔜] in np.column_stack((set𝔛, set𝔜)):
    print(f"    {str(𝔛):30} --> {𝔜}")

# Plot result
plt.figure("Simulation of function set representation")
plt.grid(color="gray")
plt.plot([𝔛.t for 𝔛 in set𝔛], [𝔜.ω_1 for 𝔜 in set𝔜], "g", label="ω_1")
plt.plot([𝔛.t for 𝔛 in set𝔛], [𝔜.ω_2 for 𝔜 in set𝔜], "r", label="ω_2")
plt.show()
